### DISCUSSION TITLE
How do I keep up to date with the changes on this repository?

### DISCUSSION THREAD

**Discussion by bassamanator** [bassamanator/Sovol-SV06-firmware#37] [Announcements]
\*\* _Flashing the firmware is not required for any method_

## Should You Update?

I treat this config like software, and like any piece of modern software, it will require periodic updates to fix issues, add features, improve functionality, etc. So, if you want to have the latest and greatest, I recommend you periodically pull the changes from this repo.

### What Changes Should You Be Concerned With

The only changes you need to be concerned with relate to the `.cfg` and `.conf` files; all changes to `.md` files (these are markdown files, the text that you see when you visit this repo) can be ignored.

### Git Users

If you do a `git diff origin/master --name-only` and it shows you that a `.cfg` file has changed, you _might_ want to pull in those changes.

### Non-Git Users

Unfortunately, there's no simple way for you to ascertain whether there have been changes. I simply recommend that you update your config via the instructions below from time to time.

## How to Update

There are three methods to accomplish this:

1. The correct method, using `git`.
2. The mostly acceptable method, without using `git`.
3. The start-from-scratch method. _Recommended for non-git users and those who haven't made many changes to their configuration._

### The `git` method.

⚠️ This method is only intended for those who already know how to use `git`.

📝 In this example, I'm using the `master` branch which applies to SV06 users only. Be sure to use the appropriate branch for your printer.

📝 The assumption is that you didn't change branches after the initial setup, so you are in `master`, and there are many untracked changes.

1. `ssh` into your Klipper host.
2. `cd ~/printer_data/config`
3. `git checkout -b my-settings`
4. `git add .`
5. `git commit -m "Saving my settings."`
6. `git checkout master`
7. `git pull`
8. `git checkout my-settings`
9. `git merge master --no-ff --no-commit`
10. Deal with comparing and pulling in incoming changes from `master` using your favourite code editor. You can also do this from the command line via `git commit --interactive`, however, only advanced `git` users should attempt this, though it is easy enough.
11. Add all welcomed changes to the staging area, and discard all other changes.
12. `git commit -m "Update with upstream."`

You should push your branch to your own fork of this repo.

For any future updates, you can run through the same process again, however, you cannot re-create the `my-settings` branch as you did in `step 3`, because it already exists. Simply omit the `-b` flag in `step 3` next time you update.

### The mostly acceptable method.

This method has shortcomings, because it relies on the user's memory, and requires more manual edits. Perfectly functional method, however.

1. Read all the documentation.
2. Backup your current configuration, essentially everything inside `~/printer_data/config`.
3. Repeat steps in [Download OSS Klipper Configuration](https://github.com/bassamanator/Sovol-SV06-firmware#download-oss-klipper-configuration), and [Adjust Configuration with MCU Path](https://github.com/bassamanator/Sovol-SV06-firmware#adjust-configuration-with-mcu-path).
4. Copy everything from `#*# <---------------------- SAVE_CONFIG ---------------------->` onward and paste into new `printer.cfg`, inclusive.
5. Copy any other changes you might have made into the new configuration. Maybe you had adjusted the size of your printer (`position_max`), or other such changes.

### The start-from-scratch method.

1. Delete the folder `~/printer_data/config`.
2. Recreate the folder `~/printer_data/config` via `mkdir ~/printer_data/config`.
3. Start the Klipper installation process starting from `Download OSS Klipper Configuration` in the appropriate branch.

## Linux Tips

- In linux, you can delete files via `rm fileName` and directories via `rmdir directoryName`.
- In linux, you can list files and folders via `ls -lah`.

You are now up-to-date with this repo, and have added your personal settings on top.

**Comment by muffix** [comment]
Another option to stay up to date would be to have this repo managed by the Moonraker update manager. It could then be updated from the web interface, just like Klipper and Moonraker itself.

To do that, you can add the following to the `moonraker.conf`: 
```
[update_manager printer_config]
type: git_repo
primary_branch: master
path: ~/printer_data/config
origin: https://github.com/bassamanator/Sovol-SV06-firmware.git
managed_services: klipper
```

The only caveat is that it expects no local changes to the repository; otherwise it refuses to update.

Assuming the only changes made are to the `printer.cfg` (e.g. by adding a bed mesh or other calibrations), we can make git assume that there are no changes to the file by setting `git update-index --skip-worktree printer.cfg`. This allows us to update from the web interface. It would, however, break if upstream changes to the `printer.cfg` are made.

**Comment by ChrSch84** [comment]
Could someone explane a noob like me how to update my pi 3 with the new git? Thank you

**Reply by bassamanator** [reply]
Updating the Rpi3 is a different matter, you're updating the OS, not the config. But here's the command: `sudo apt update && sudo apt full-upgrade`. Klipper will likely stop working because of recent changes in Debian. You'll have to find out what port your `mcu` is connected to via the `by-path` method mentioned in the [Initial Steps](https://github.com/bassamanator/Sovol-SV06-firmware#initial-steps), and then make adjustments to your `printer.cfg` accordingly.

**Reply by ChrSch84** [reply]
Sry this is a missunderstanding. I mean this part:

"git checkout -b my-settings
git add .
git commit -m "Saving my settings."
git checkout master
git pull
git checkout my-settings
git merge master --no-ff --no-commit"

Where do i put this? on the Pi? On the Github Page?

Is there no easy way to update to your latest Version? Thanks.

**Reply by RandomSpaceship** [reply]
Each line of that is a command that you'll need to run in the terminal (in that order), which you should have previously connected to using ssh. Git is the program that manages code versions and pulls the latest code from this repository - it's very well documented if you want to look it up. What the above commands effectively do is save your (local) changes, then _merge_ the updates into them.

**Reply by ChrSch84** [reply]
Ah ok thank you. So not realy noob Friendly :( 

Maybe someone could make a guide for this.

**Reply by bassamanator** [reply]
@ChrSch84 
>Is there no easy way to update to your latest Version? Thanks.

My friend, I've listed 2 methods on how to update, each including all steps necessary to update. The n00b friendly method requires you to know how to use a text editor, that is all. How much easier can I make this?

**Reply by ChrSch84** [reply]
@bassamanator thanks for the 2 methods. I use now method 2 because method 1 does not work for me. Even the first command give me a error.
"$ git checkout -b my-settings
fatal: not a git repository (or any of the parent directories): .git"
Thanks for the Help

**Reply by bassamanator** [reply]
@ChrSch84 I'm glad you are able to keep up with the repo.

On a different note, I've been working on a git reference/tutorial for beginners. You can find it [here](https://github.com/bassamanator/git-for-beginners), it's far from complete, however. Git is too difficult a tool to learn just for the sake of keeping up with 1 repository, but if ever get into any kind of coding, it's a must learn. Lot of great Youtube tutorials out there as well. Best!

**Comment by gregma** [comment]
Just something very minor.  "Hot to Update" should be "How to Update"

**Comment by dasb00ter** [comment]
Hey this may be a stupid question but it seems to me there are two different portions to klipper. The firmware flashed onto the printer and then the controller part of klipper whether it be a Pi or btpad 7 or whatever. Should I also be updating firmware along with the controller portion of Klipper? You are also right about the memory thing(in your instructions) here I cannot exactly remember what changes I have made to printer.cfg and I am quite sure I have :- ( I will have to parse the file by looking sadly.

**Reply by bassamanator** [reply]
There is no need to flash the board with new firmware, but you can if you want to. 

You should simply be updating Klipper via the dashboard.

**Reply by dasb00ter** [reply]
I wouldn't unless it's advantageous in terms of some new feature or something is broken. 

So for example x twist compensation would that have required at some point an update in firmware? Or would this have been implemented purely on the dashboard side of things?

**Reply by bassamanator** [reply]
If you try to use a feature that requires a firmware update (because you've added a section to your `printer.cfg` that is only available in the latest Klipper release, for example), you will be warned in the dashboard that a firmware update is required. This is a very rare occurrence.
